クラス(Javascript)の課題提出

yokomaru opened the pull request:
# 課題URL
https://bootcamp.fjord.jp/products/21950

# 手順など
https://github.com/yokomaru/js-practices/blob/21eaf6f26d4f7075871fac3ebda618e9aadf5f57/04.class/readme.md

yokomaru committed:
add: install required npms

- 必要なnpmをインストール
    - minimist
    - sqlite3
    - readline
    - enquirer

yokomaru committed:
add: files

必要なファイルを追加

yokomaru committed:
add: init.js

- 初回にDBとテーブルを作成するスクリプト

yokomaru committed:
add: index.js

- 各クラスを定義する
- 処理の実行(App.main)

yokomaru committed:
add: App.js

- オプションによって分岐して各コマンドを実行する

yokomaru committed:
add: command.js

- 各コマンドの諸々の処理を担当する

yokomaru committed:
add: memo-control.js

- 各メモの操作(index, delete, crette)を担当する
- DBをここでは直接扱わず、Data BaseOperationに必要な値を渡しその結果を元に処理を行う

yokomaru committed:
add: database-operation.js

- sqliteの処理をプロミスにラップし、データベースへの処理を担当する

yokomaru committed:
add: memo.js

- メモに関するデータを保持するクラス

yokomaru committed:
add: oprion.js

- コマンドオプションを保持するクラス

yokomaru committed:
add: select-prompt.js

- 選択表示を担当するクラス

yokomaru committed:
add: user-input-interface.js

- メモ入力を担当するクラス
- user-input(ユーザー入力)というクラス名だと分かりづらく感じたためuser-input-interfaceに変更

yokomaru committed:
add: shebang

- shebangを追加＆ファイルの権限を変更してコマンドのように実行できるようにする

yokomaru committed:
add: readme

- 初回実行時の説明をするreadme.mdを追加

yokomaru committed:
fix: readme

- 誤字を修正
- メモアプリの操作の項目を追加

yokomaru committed:
fix: select sql

- アスタリスクでカラムを全選択していたが、必要なカラムに絞って表示する

yokomaru committed:
fix:  name in memo to split on new line instead of slice

- 改行が見当たらない時に-1を返すため、最後の一文字が削除されてしまうので改行文字でsplitし最初の要素を返すように修正

yokomaru committed:
Update readme.md

gifの追加

yokomaru committed:
fix: method order in alphabetical order

yokomaru committed:
add:  new line

`npm run fix`を実行した結果を反映

yokomaru committed:
fix: add # to private fields

- つけ忘れていたのでprivate fieldに#を追加

yoshitsugu commented on 04.class/lib/memo.js:
```diff
@@ -0,0 +1,14 @@
+
+  #generateDisplayName = () => {
+    const slicedName = this.content?.split("\n")[0];
+    return !slicedName ? "No Title" : slicedName;
```
これだと"No Title"という名前のメモが作られた場合との区別がつかなさそうです。
1行目がないということはメモ自体ない、ということに思うので、(改行文字からはじまる、というのは通常あまり考えられないと思います)そういうメモは作られないようにしたほうがよいのではと思いました。

yoshitsugu commented on 04.class/lib/memo.js:
```diff
@@ -0,0 +1,14 @@
+
+  #generateDisplayName = () => {
+    const slicedName = this.content?.split("\n")[0];
+    return !slicedName ? "No Title" : slicedName;
```
これだと"No Title"という名前のメモが作られた場合との区別がつかなさそうです。
1行目がないということはメモ自体ない、ということに思うので、(改行文字からはじまる、というのは通常あまり考えられないと思います)そういうメモは作られないようにしたほうがよいのではと思いました

yokomaru replied:
1行目がない(改行のみ、または全て空白やスペースなどの文字)場合はメモが作られないように修正いたしました！

2b291fa384b539ee3e30bb19e3666ad9ccf1777b
b4f0b4145e66978ad570a270ee241159421af0c5

yoshitsugu commented on 04.class/lib/select-prompt.js:
```diff
@@ -0,0 +1,34 @@
+import Enquirer from "enquirer";
+
+class SelectPrompt {
```
あまりこれとUserInputInterfaceをわける必要性を感じませんでした。
同じ標準入力をあつかうクラスとして統一できないでしょうか。

yoshitsugu commented on 04.class/lib/option.js:
```diff
@@ -0,0 +1,14 @@
+class Option {
```
これも結局標準入力に依存してしまっている (argv依存になっている)ので、標準入力依存の部分はまとめたほうが、たとえば入力をWeb APIに変更するような場合も修正しやすくなると思います。

yokomaru replied:
https://github.com/yokomaru/js-practices/pull/4#discussion_r2070918807 で指摘いただいたものと合わせて以下で対応いたしました！

[1e1817b](https://github.com/yokomaru/js-practices/pull/4/commits/1e1817b20267f9258aab868c5a1b23389f8cb31f)

yoshitsugu reviewed (commented):
クラスが複雑で分ける必要性が直感的にわかりづらいので、たとえば、

- メモの核となるドメインロジックを扱う部分(今回そんなにロジックないですが。。)
- 永続化部分(今回はDBを使っていますね)
- ユーザーとのやりとり部分(今回は標準入出力を使っていますね)
- それらを統合したもの

くらいでわけてしまってもよいのかなと思います。

yokomaru committed:
fix: Prevent creation of notes with blank first lines or completely blank notes.

- 入力時
    - メモの1行目が空行の場合はメモの配列にpushしない
    - メモの配列が全てない場合はrejectを返しメモが作成できないようにする

yokomaru committed:
fix: Consolidate user input-related functions into a single class

- クラスが複雑になっているため、ユーザーとのやりとり部分をUserInputクラスにまとめる
    - SelectPrompt
    - UserInterface
    - Option

- まとめた上で、CommandのプロパティとしてUserInputを持たせるように修正

yokomaru committed:
fix: Prevent notes from being created not only when the first line is blank, but also when there are blank characters or tabs

- 現状だと空白文字やタブが入っていると文字が入力されていると判断されてしまい登録できてしまうため、trimをかけて""に変換できてしまったものは最初の1行目には登録できないようにする(2行目以降は可能)
- 等価演算を厳密にする

yokomaru committed:
fix: Make the equivalence operator a strict equivalence operator.

- 意図しない比較結果を避けるために厳密等価演算子を使用する

yokomaru committed:
fix: readme

- 1行目の制限について追加

yoshitsugu commented on 04.class/lib/app.js:
```diff
@@ -0,0 +1,27 @@
+  static async main(command) {
+    if (command.userInput.hasNoOption()) {
+      await command.createMemo();
+      process.exit(1);
```
ステータスコード1は一般的にはエラー終了をあらわします。正常終了としたい場合は0にしましょう。

yoshitsugu reviewed (approved):
OKです！
